nginx/lua/redirect_content.lua

```diff
@@ -0,0 +1,96 @@
+local ngx = require("ngx")
+local resty_http = require("resty.http")
+local fileutil = require("fileutil")
+local config = require("config")
+local gossip = require("gossip")
+
+
+-- TODO: always "redirect"?
+local path = ngx.var.request_uri:sub(#"/redirect" + 1)
+local webdav_uri = config.data.uriprefix .. path
+local stat = fileutil.get_metadata(config.data.local_path .. path, false)
+if stat.exists then
+    ngx.status = ngx.HTTP_TEMPORARY_REDIRECT
+    -- TODO: do we want to use the full url (config.data.server_address)
+    ngx.header["Location"] = webdav_uri
+    ngx.exit(ngx.OK)
+end
+
+---@type function
+---@param peer string Peer hostname
+---@param uri string File url to check
+---@param token string Bearer token for authentication
+---@return {peer: string, location: string?} res Response from peer
+local function peer_query_file(peer, uri, token)
+    local httpc = resty_http.new()
+    -- peer has trailing slash, uri has leading slash
+    local location = peer .. uri:sub(2)
+    ngx.log(ngx.NOTICE, "Querying location: ", location)
+    local res, err = httpc:request_uri(location, {
+        method = "HEAD",
+        headers = {
+            ["Authorization"] = "Bearer " .. token,
+            ["User-Agent"] = "nginx-webdav/" .. config.data.server_version,
+        },
+    })
+    if not res then
+        ngx.log(ngx.ERR, "Failed to send file query to " .. peer .. ": " .. err)
+        -- TODO: update peer status for gossip (unreachable)
```

Comment:
We should update our peer list when one is unreachable or if it has very poor redirect performance (e.g. latency)